src/frontend/core/executor/impl/TriangleCountExecutor.cpp

```diff
@@ -34,13 +50,400 @@ TriangleCountExecutor::TriangleCountExecutor(SQLiteDBInterface *db, PerformanceS
+            p_avail.clear();
+            alloc.insert(alloc2.begin(), alloc2.end());
+            loads.insert(loads2.begin(), loads2.end());
+            return 0;
+        }
+        if (rem < best_rem) {
+            best_rem = rem;
+            best = {.alloc = alloc2, .remain = remain2, .p_avail = p_avail2, .loads = loads2};
+        }
+    }
+    alloc.insert(best.alloc.begin(), best.alloc.end());
+    remain.clear();
+    remain.insert(best.remain.begin(), best.remain.end());
+    p_avail.clear();
+    p_avail.insert(best.p_avail.begin(), best.p_avail.end());
+    loads.clear();
+    loads.insert(best.loads.begin(), best.loads.end());
+    return best_rem;
+}
+
+static std::vector<int> reallocate_parts(std::map<int, string> &alloc, std::set<int> &remain,
+                                         const std::map<int, std::vector<string>> &P_AVAIL) {
+    map<int, int> P_COUNT;
+    for (auto it = P_AVAIL.begin(); it != P_AVAIL.end(); it++) {
+        P_COUNT[it->first] = it->second.size();
+    }
+    vector<int> remain_l(remain.begin(), remain.end());
+    sort(remain_l.begin(), remain_l.end(),
+         [&P_COUNT](int &p1, int &p2) { return P_COUNT[p1] > P_COUNT[p2]; });  // partitions with more copies goes first
+    vector<int> PARTITIONS;
+    for (auto it = P_COUNT.begin(); it != P_COUNT.end(); it++) {
+        PARTITIONS.push_back(it->first);
+    }
+    sort(PARTITIONS.begin(), PARTITIONS.end(), [&P_COUNT](int &p1, int &p2) {
+        return P_COUNT[p1] < P_COUNT[p2];
+    });  // partitions with fewer copies goes first
+    vector<int> copying;
+    while (!remain_l.empty()) {
+        int p0 = remain_l.back();
+        remain_l.pop_back();
+        int w_cnt = P_COUNT[p0];
+        if (w_cnt == 1) {
+            copying.push_back(p0);
+            continue;
+        }
+        const auto &ws = P_AVAIL.find(p0)->second;
+        bool need_pushing = true;
+        for (auto it = PARTITIONS.begin(); it != PARTITIONS.end(); it++) {
+            int p = *it;
+            if (w_cnt <= P_COUNT[p]) {
+                copying.push_back(p0);  // assuming PARTITIONS are in sorted order of copy count
+                need_pushing = false;
+                break;
+            }
+            if (alloc.find(p) == alloc.end()) {
+                continue;
+            }
+            auto w = alloc[p];
+            if (std::find(ws.begin(), ws.end(), w) != ws.end()) {
+                alloc.erase(p);
+                alloc[p0] = w;
+                remain_l.push_back(p);
+                need_pushing = false;
+                break;
+            }
+        }
+        if (need_pushing) copying.push_back(p0);
+    }
+    return copying;
+}
+
+static void scale_up(std::map<string, int> &loads, map<string, string> &workers, int copy_cnt) {
+    int curr_load = 0;
+    for (auto it = loads.begin(); it != loads.end(); it++) {
+        curr_load += it->second;
+    }
+    int n_cores = copy_cnt + curr_load - 3 * loads.size();
```

Comment:
What is the rationale behind this equation? Why we have 3 there?

Reply:
3 is the maximum number of tasks we allocate to a worker under normal conditions. This formula gives the number of additional cores we need to handle the new load.

Reply:
How did we come up with the value 3? why not 4/5/10 etc.? What do you mean by normal conditions? then what are the abnormal conditions?

Reply:
3 = 4 - 1  where 4 is the number of cores available in each machine of the cluster.<br>
normal conditions are when the load can be handled by scaling out (the cluster has not hit the maximum worker limit)